### ISSUE TITLE
Dealing with proliferation of package variants

### ISSUE THREAD

**Issue by h-vetinari** [conda-forge/conda-forge.github.io#1558]
Crossposting from https://github.com/conda/conda/issues/11053 because it affects conda-forge in many ways.

> More and more packages come in various different flavours, independently of relatively fixed things like the python version. A non-exhaustive list:
> - cpu vs cuda (vs. eventually ROCm or Metal)?
> - blas-flavour
> - mpi-flavour
> - openmp-enabled or not
> - possibly for CPU [arches](https://github.com/conda-forge/conda-forge.github.io/issues/1261)

I opened that issue on conda to find out if there might be a better solution than tacking on more and more things onto buildstrings. 😅

**Comment by jakirkham** [comment]
Yes virtual packages. Particular custom ones for extensibility ( https://github.com/conda/conda/issues/10131 ). They can detect system features, which either allow or disallow particular variants.

**Comment by h-vetinari** [comment]
Virtual packages are not the answer to everything though... A virtual package IMO makes sense for something (effectively) immutable like OSX, CPU Arch, GPU _version_, but not for things like MPI, OpenMP, etc... And even with virtual packages, I might want to override them and install the CPU version despite having a GPU, or install the SSE version despite being AVX2-capable.

**Comment by h-vetinari** [comment]
Though looking closer at https://github.com/conda/conda/issues/10131, maybe I'm overlooking how the plugin-approach might overcome these issues/limitations? If so, could you elaborate?

**Comment by jakirkham** [comment]
Mainly thinking about auto-detection and constraint of OS and hardware features. With the existing virtual packages there are already ways to override these. Presumably that could be done for new ones

The other things (BLAS, MPI, etc.) are up to the user to decide. What alternative would you propose?

**Comment by h-vetinari** [comment]
> The other things (BLAS, MPI, etc.) are up to the user to decide. What alternative would you propose?

My point was exactly about exposing these variants to the user, and ideally in a more user-friendly way than pattern-matching build strings. I have nothing against package variants, I think they're a great asset of conda-forge. I just wish we'd have a better UI for it.

**Comment by ngam** [comment]
@h-vetinari what do you think about splitting cuda arches into separate ones, so instead of cuda 11.2, we will have `sm`s: 
- 3.5;
- 5.0+PTX;
- 6.0;
- 6.1;
- 7.0;
- 7.5;
- 8.0; and 
- 8.6

xref: https://github.com/conda-forge/pytorch-cpu-feedstock/issues/108 (comment: https://github.com/conda-forge/pytorch-cpu-feedstock/issues/108#issuecomment-1144162948)

**Comment by jakirkham** [comment]
@ngam, this might be best as a new issue on the webpage (this) repo (though it does seem like an interesting discussion to have)

**Comment by ngam** [comment]
> @ngam, this might be best as a new issue on the webpage (this) repo (though it does seem like an interesting discussion to have)

I don't follow. My speculative question is not related to the above discussion? I am not really sure about this idea anyway, so I just wanted to get h-v thoughts. I can delete the comment if you want, or feel free to transfer it somewhere else

**Comment by jakirkham** [comment]
It's diving deeper into CUDA functionality. We may want to invite others familiar with that use case to discuss further. While generally within this issue, it is a bit of a subcase. So we may not want to get too into the weeds on that case in this issue (and so move it to another one)

To be clear am happy to see that discussion move forward, just trying to keep things organized

**Comment by jakirkham** [comment]
FWIW the virtual package plugin architecture is being worked on in PR ( https://github.com/conda/conda/pull/11854 ). Once complete virtual packages could be created independently and used as dependencies in metapackages to cutdown on the configuration users themselves would have to do.
